File: app/src/main/java/com/stedi/multitouchpaint/view/CanvasView.java
package com.stedi.multitouchpaint.view;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import com.stedi.multitouchpaint.data.Brush;
import com.stedi.multitouchpaint.painter.BasePainter;

public class CanvasView extends View {
    private BasePainter painter;
    private Brush brush;

    public CanvasView(Context context) {
        this(context, null);
    }

    public CanvasView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CanvasView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setBackgroundColor(Color.WHITE);
    }

    public void setPainter(BasePainter painter) {
        this.painter = painter;
        invalidate();
    }

    public BasePainter getPainter() {
        return painter;
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
            case MotionEvent.ACTION_POINTER_DOWN: {
                painter.onPointerDown(event, brush, this);
                break;
            }

            case MotionEvent.ACTION_MOVE: {
                painter.onPointerMove(event, brush, this);
                break;
            }

            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_POINTER_UP: {
                painter.onPointerUp(event, brush, this);
                break;
            }

            default:
                return false;
        }
        return true;
    }

    @Override
    protected void onDraw(Canvas viewCanvas) {
        painter.onDraw(viewCanvas);
    }

    public void setBrush(Brush brush) {
        this.brush = brush;
    }

    public void undo() {
        if (painter.onUndo())
            invalidate();
    }

    public Bitmap generatePicture() {
        Bitmap bitmap = Bitmap.createBitmap(getWidth(), getHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        draw(canvas);
        return bitmap;
    }

    public void clearPicture() {
        if (painter.onClear())
            invalidate();
    }

    public void setPicture(Bitmap bitmap) {
        clearPicture();
        painter.onSetPicture(bitmap, getWidth(), getHeight());
        invalidate();
    }

    public boolean isDrawing() {
        return painter.isDrawing();
    }
}

File: app/src/main/java/com/stedi/multitouchpaint/view/CanvasView.kt
package com.stedi.multitouchpaint.view

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import com.stedi.multitouchpaint.data.Brush
import com.stedi.multitouchpaint.painters.Painter

class CanvasView : View {
    private lateinit var painter: Painter
    private lateinit var brush: Brush

    constructor(context: Context?) : this(context, null)
    constructor(context: Context?, attrs: AttributeSet?) : this(context, attrs, 0)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        setBackgroundColor(Color.WHITE)
    }

    fun setPainter(painter: Painter) {
        if (this::painter.isInitialized) {
            this.painter.onDetach(this)
        }
        this.painter = painter
        this.painter.onAttach(this)
        invalidate()
    }

    fun getPainter() = painter

    fun setBrush(brush: Brush) {
        this.brush = brush
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        return when (event.actionMasked) {
            MotionEvent.ACTION_DOWN, MotionEvent.ACTION_POINTER_DOWN -> {
                painter.onPointerDown(event, brush, this)
                true
            }
            MotionEvent.ACTION_MOVE -> {
                painter.onPointerMove(event, brush, this)
                true
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_POINTER_UP -> {
                painter.onPointerUp(event, brush, this)
                true
            }
            else -> false
        }
    }

    override fun onDraw(canvas: Canvas) {
        painter.onDraw(canvas)
    }

    fun undo() {
        if (painter.onUndo()) {
            invalidate()
        }
    }

    fun clearPicture() {
        if (painter.onClear()) {
            invalidate()
        }
    }

    fun isDrawing() = painter.isDrawing

    fun generatePicture(): Bitmap {
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        draw(Canvas(bitmap))
        return bitmap
    }

    fun setPicture(bitmap: Bitmap) {
        clearPicture()
        painter.onSetPicture(bitmap, width, height)
        invalidate()
    }
}